src/libbson/tests/test-bson-vector.c

```diff
@@ -0,0 +1,1400 @@
+   {
+      bson_t converted = BSON_INITIALIZER;
+      bson_iter_t iter;
+      ASSERT (bson_iter_init_find (&iter, &doc, "longer_vector"));
+      ASSERT (BSON_APPEND_ARRAY_FROM_VECTOR (&converted, "array", &iter));
+      ASSERT (bson_iter_init_find (&iter, &converted, "array"));
+      ASSERT (BSON_ITER_HOLDS_ARRAY (&iter));
+      ASSERT (bson_iter_recurse (&iter, &iter));
+      ASSERT (BSON_APPEND_VECTOR_PACKED_BITS_FROM_ARRAY (&doc, "longer_round_trip", &iter, NULL));
+      bson_destroy (&converted);
+   }
+   {
+      bson_iter_t a, b;
+      ASSERT (bson_iter_init_find (&a, &doc, "longer_vector"));
+      ASSERT (bson_iter_init_find (&b, &doc, "longer_round_trip"));
+      ASSERT (bson_iter_binary_equal (&a, &b));
+   }
+
+   // Padding bits will be initialized to zero when a packed_bits vector is first allocated by
+   // bson_append_vector_packed_bits
+   {
+      // Set the uninitialized part of 'doc' to a known value
+      static const uint32_t reserve_len = 512;
+      memset (bson_reserve_buffer (&doc, doc.len + reserve_len) + doc.len, 0xdd, reserve_len);
+
+      bson_vector_packed_bits_view_t view;
+      ASSERT (BSON_APPEND_VECTOR_PACKED_BITS (&doc, "padding_init_test", 12, &view));
+      ASSERT (bson_vector_packed_bits_view_length_bytes (view) == 2);
+      ASSERT (bson_vector_packed_bits_view_padding (view) == 4);
+
+      // BSON validity only requires the low 4 bits to be zero, but the entire last
+      // byte will be zeroed by our implementation.
+      uint8_t bytes[2];
+      ASSERT (bson_vector_packed_bits_view_read_packed (view, bytes, sizeof bytes, 0));
+      ASSERT_CMPUINT ((unsigned) bytes[0], ==, 0xdd);
+      ASSERT_CMPUINT ((unsigned) bytes[1], ==, 0x00);
+   }
+
+   // Padding bits can't be forcibly given nonzero values using bson_vector_packed_bits_view_write_packed
+   {
+      bson_vector_packed_bits_view_t view;
+      ASSERT (BSON_APPEND_VECTOR_PACKED_BITS (&doc, "padding_mask_test", 13, &view));
+      ASSERT (bson_vector_packed_bits_view_length_bytes (view) == 2);
+      ASSERT (bson_vector_packed_bits_view_padding (view) == 3);
+
+      uint8_t bytes[2] = {0xff, 0xff};
+      ASSERT (bson_vector_packed_bits_view_write_packed (view, bytes, sizeof bytes, 0));
+      ASSERT (bson_vector_packed_bits_view_read_packed (view, bytes, sizeof bytes, 0));
+      ASSERT_CMPUINT ((unsigned) bytes[0], ==, 0xff);
+      ASSERT_CMPUINT ((unsigned) bytes[1], ==, 0xf8);
+   }
+
+   bson_destroy (&doc);
+}
+
+#define MAX_TESTED_VECTOR_LENGTH 10000
+#define FUZZ_TEST_ITERS 5000
+
+static void
+test_bson_vector_view_api_fuzz_int8 (void)
+{
+   size_t current_length = 0;
+   bson_t vector_doc = BSON_INITIALIZER;
+   int8_t *expected_elements = bson_malloc (MAX_TESTED_VECTOR_LENGTH * sizeof *expected_elements);
+   int8_t *actual_elements = bson_malloc (MAX_TESTED_VECTOR_LENGTH * sizeof *actual_elements);
+   for (int fuzz_iter = 0; fuzz_iter < FUZZ_TEST_ITERS; fuzz_iter++) {
+      int r = rand ();
```

Comment:
```suggestion
      size_t r = _mongoc_simple_rand_size_t ();
```
Recommend using the `_mongoc_simple_rand_*` functions provided by `mongoc-util-private.h` for better type consistency and to avoid increasing our raw `rand()` footprint.

Suggestion also applies to other functions below.

Reply:
The rand functions definitely leave a lot to be desired w.r.t. performance. Their addition in https://github.com/mongodb/mongo-c-driver/pull/898 was very much a "correctness first, performance later" decision. The suggestion to use them here was only for `rand()`-avoidance and reuse of existing API to easily achieve `size_t` type consistency without requiring explicit casts (i.e. in module expressions). imo given the execution of these tests are still under a second in total, I do not think the performance tradeoff is especially significant given the overall execution time of the test suite. If you still prefer the explicit `rand()`, I am fine with that decision.